### PR TITLE
perf: Branchless Parquet Prefiltering

### PR DESCRIPTION
**Pull request by coastalwhite**
Change the Parquet decoders to use branchless filtering for the `prefiltered` methods.

This rewrites most of the decoders.

### CODE REVIEW

File: crates/polars-parquet/src/arrow/read/deserialize/primitive/integer.rs [hunk 1]
```diff
@@ -358,6 +361,258 @@ where
...
+        out: *mut T,
+        dfn: D,
+    ) -> (usize, usize)
+    where
+        T: NativeType,
+        P: ParquetNativeType,
+        i64: num_traits::AsPrimitive<P>,
+        D: DecoderFunction<P, T>,
+    {
+        let mut num_read = 0;
```
Comment (orlp) [inline crates/polars-parquet/src/arrow/read/deserialize/primitive/integer.rs:509]: I think this loop can be written more efficiently.

```rust
let num_valid = validity.count_ones();
let mut num_written = 0;

while filter != 0 {
    let offset = filter.trailing_zeros();
    validity >>= offset;
    
    let idx = num_valid - validity.count_ones();
    let v = if validity & 1 != 0 {
        dfn.decode(unsafe { values.get_unchecked(idx) })
    } else {
        T::zeroed()
    };

    *out.add(num_written) = v;

    num_written += 1;
    filter >>= offset + 1;
    validity >>= 1;
}

(num_valid, num_written)
```

File: crates/polars-parquet/src/arrow/read/deserialize/primitive/integer.rs [hunk 2]
```diff
@@ -358,6 +361,258 @@ where
...
+
+            num_rows_done += nr;
+            unsafe { target.set_len(start_length + num_rows_done) };
+            target.resize(start_length + num_rows, T::zeroed());
+            return;
+        }
+
+        let mut v = v;
+        let mut f = f;
+        let mut num_read = 0;
```
Comment (orlp) [inline crates/polars-parquet/src/arrow/read/deserialize/primitive/integer.rs:577]: Idem.

File: crates/polars-parquet/src/arrow/read/deserialize/utils/dict_encoded.rs [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+            assert!((highest_idx as usize) < dict.len());
+
+            buffer_side = !buffer_side;
+            num_buffered += num_added_values;
+        }
+
+        // @NOTE: We have to cast to u64 here to avoid the `shr` overflow on the filter.
+        let mut f = f as u64;
+        let mut v = v as u64;
+        let mut num_read = 0;
```
Comment (orlp) [inline crates/polars-parquet/src/arrow/read/deserialize/utils/dict_encoded.rs:327]: Idem.

**Review by orlp — CHANGES_REQUESTED** [review]

File: crates/polars-parquet/src/arrow/read/deserialize/primitive/integer.rs (hunk 1 above)
Comment (orlp) [reply crates/polars-parquet/src/arrow/read/deserialize/primitive/integer.rs:509]: Even better I believe:
```rust
let num_available = validity.count_ones();
let mut num_written = 0;

while filter != 0 {
    let offset = filter.trailing_zeros();
    let idx_to_read = num_available - (validity >> offset).count_ones();
    let v = if validity & 1 != 0 {
        dfn.decode(unsafe { values.get_unchecked(idx_to_read) })
    } else {
        T::zeroed()
    };

    *out.add(num_written) = v;
    num_written += 1;
    filter &= filter - 1;
}

(num_available, num_written)
```

**Comment by coastalwhite** [comment]
ping @ritchie46.

**Pull request merged**